### DISCUSSION TITLE
How to send/receive email@example.com emails on 5 different postal servers: mail0., mail1., mail2., mail3. & mail4.example.com?

### DISCUSSION THREAD

**Discussion by JohnHammell** [postalserver/postal#3180] [Help with using Postal]
Is there a specific way required in order to setup & configure 5 (or any number) Postal servers each using a different IP address for sending & receiving email for the same root domain example.com? These also could be on different physical servers.

mail0.example.com & IP0
mail1.example.com & IP1
mail2.example.com & IP2
mail3.example.com & IP3
mail4.example.com & IP4

So email@example.com emails could be sent through any of the 5 Postal servers and emails to that address also received by any of the Postal servers.

What I'm concerned about is mainly with return path domains, bounces, etc. That they be handled properly, setup by me properly. For example, maybe a message sent via mail4.example.com and the bounce message going to mail1.example.com which did not send that email, so would have no record of it. Or would a bounce never go to any other Postal server other than the one that sent it (mail4, in this example)?

Something else I noticed in email headers is that with my current Postal server on mail.example.com there are message IDs such as <20241020002937.13d8d1cfb421@example.com> and am wondering if/how that might affect running things on multiple mail servers.

Also, in the Postal DNS setup for example.com, there's an optional but recommended (to improve deliverability) setting of pointing a psrp.example.com CNAME to rp.mail.example.com. How could that work in this type of setup where example.com sends/receives on multiple Postal servers?

Anything else I might be overlooking?

**Comment by willpower232** [comment]
If you're intending on using the same pile of databases for all your Postal servers then it would work roughly the same way as you describe.

The solution would be to have multiple copies of DNS records for mail.example.com rather than mail0-5.example.com (I think they call it round-robin DNS but I'm not sure). Each server would have identical configuration values.

That way any of the servers could handle incoming or bounced messages

**Comment by JohnHammell** [comment]
My intention was to use each as an independent Postal server, installed via the regular directions.

Sounds like you're suggesting that the different Postal servers mail0-4 all share the same database or databases. That might be something I could consider, seems it may be a good option. Do you know of any Postal docs with directions on how to setup in that way?

**Reply by willpower232** [reply]
Postal is fully clusterable so as long as you point all your copies to the same database, you should be able to use it wherever you wish, I don't believe there are separate docs, it should "just work". I think the most annoying part would be you will have to ensure the SPF record actually correlates to all 5 servers.

If you keep them as separate piles of databases, you would have to log in to all 5 manually to figure out where a message was eventually sent from and what happened to us unless you're able to set up webhooks to receive this information back.

> So email@example.com emails could be sent through any of the 5 Postal servers and emails to that address also received by any of the Postal servers.

If you're manually adding the same domain to each one, you may want to consider setting up one and then copying the database across to reduce the number of DKIM records you have to manage.

> What I'm concerned about is mainly with return path domains, bounces, etc. That they be handled properly, setup by me properly. For example, maybe a message sent via mail4.example.com and the bounce message going to mail1.example.com which did not send that email, so would have no record of it. Or would a bounce never go to any other Postal server other than the one that sent it (mail4, in this example)?

If you keep them all as separate then you would certainly have a weird time with the return path. Your psrp.example.com would have to refer rp0-4.example.com and if a Postal received a bounce for a message it did not send then I think you would need a webhook to actually understand that your message bounced and was not successfully delivered, the information contained separately in each Postal would not be able to be understood fully on its own.

> Something else I noticed in email headers is that with my current Postal server on mail.example.com there are message IDs such as 20241020002937.13d8d1cfb421@example.com) and am wondering if/how that might affect running things on multiple mail servers.

In theory these values are supposed to be unique enough that it isn't a problem, I don't think you would have a problem sending from the same domain in Postal and Outlook for example.

**Reply by JohnHammell** [reply]
I'm exploring & should sometime today be testing a clustered setup as you describe. It does sound like it would be the best option versus having each mail server running its own db.

Some concerns I have regarding all 5 mail servers sharing the same db: Is the database used for storing / referencing any of the details about the individual mail servers that would not be shared / which would be unique per mail server? Such as: mail#.example.com , the unique IP addresses used by each mail server or anything else that should be unique per server.

Would the individual Postal servers get those details solely from the config file at /opt/postal/config/postal.yml on each server?

Concern here would be if, for example, the database stores the server IP then ends up using that one server IP for all 5 mail servers (which are supposed to each have a unique IP, not shared IP).

**Reply by JohnHammell** [reply]
In the config YAML file I see two different database configs, "main_db" and "message_db".

Do both of those databases need to be shared among the 5 postal servers in order to cluster them? Or does only the message_db need to be shared?

**Reply by willpower232** [reply]
> Some concerns I have regarding all 5 mail servers sharing the same db: Is the database used for storing / referencing any of the details about the individual mail servers that would not be shared / which would be unique per mail server? Such as: mail#.example.com , the unique IP addresses used by each mail server or anything else that should be unique per server.

If you're hoping to use IP Pools then I think it would get confused yes. Otherwise, it should be fine.

Technically Postal doesn't usually care about which IP address it uses. If you need to tell it on a per server basis to use a particular IP address in the config file then that should be fine.

> Do both of those databases need to be shared among the 5 postal servers in order to cluster them? Or does only the message_db need to be shared?

If they aren't consistent between the servers then it will probably break trying to re use a database number or go looking for a database which isn't there.

**Reply by JohnHammell** [reply]
Thank you for all the help. Very much appreciated.

I'll test things out with both main_db & message_db pointing to a shared mariadb install.

Not using IP pools. Each Postal server instance will have a unique IP.

**Comment by JohnHammell** [comment]
I think I may have partly figured out my concern with the rp.mail.example.com & psrp.example.com (optional but recommended by Postal) return path domains.

In the configs at https://github.com/postalserver/postal/blob/main/doc/config/yaml.yml

rp.mail. would be the actual return path domain. psrp. is the default custom_return_path_prefix setting.

After looking more at those configs, I believe using the below for each of the 5 numbered mail servers would get both the return path & custom return path pointing to the correct mail server for any messages sent by that server.
> rp.mail#.example.com
> psrp.mail#.example.com

This appeared to be an issue with using multiple mail servers since I saw "psrp.example.com" in the mail headers for one of my other sites (which uses only 1 postal server), but I would need that to reference uniquely to each mail server. Looks like changing psrp. to psrp.mail#. for the custom_return_path_prefix setting may accomplish that.

There would still be the issue of message-ID@example.com emails in the headers , but using a shared db should sort that out. Not sure if that part can be addressed with separate db mail servers without only using a unique subdomain on each mail server.

* Note: This message isn't so much a question but rather just some notes on things I believe I partly figured out, for reference for anyone who stumbles upon this discussion in the future, maybe something here will also help them. But please do feel free to reply to this comment if there's anything I got wrong here or if anyone has anything to add.

**Reply by willpower232** [reply]
> rp.mail. would be the actual return path domain. psrp. is the default custom_return_path_prefix setting.

Yeah, you would have 5 A records for rp.mail.example.com and that way it doesn't matter which IP address DNS chooses because they all work.

You would not have rp.mail1-5.example.com, only rp.mail.example.com
